### ISSUE TITLE
Exception in worker results in an error in the server.

### ISSUE THREAD

**Issue by fake-name** [celery/celery#2441]
I'm working on a system that uses celery.

In any event, I have a fairly simple setup. Client script:

```
import celery
import sys

mqServer = 'amqp://<snip>'

# This has to be global, there's some annoying magical 'findapp' stuff going on.
app = celery.Celery('tasks',
		backend = mqServer,
		broker  = mqServer)

app.conf.update(
	CELERY_TASK_SERIALIZER='json',
	CELERY_ACCEPT_CONTENT=['json'],  # Ignore other content
	CELERY_RESULT_SERIALIZER='json',
	CELERYD_PREFETCH_MULTIPLIER = 1,
	CELERY_MAX_CACHED_RESULTS = 1,      # Disable caching of results (otherwise we'll leak LOTS of memory)
)



@celery.task
class FetchTask(celery.Task):
	def __init__(self):
		print("Task Constructor")
		self.a = 0

	def run(self, x):
		print("x: '%s', self.a: '%s'" % (x, self.a))
		self.a += x
		if self.a > 1000:
			self.a = 0
			raise ValueError
		return self.a

if __name__ == '__main__':
	if len(sys.argv) == 2:
		num = sys.argv[1]
	else:
		num = 1

	# Register the FetchTask
	celery.registry.tasks[FetchTask.name]

	# Start celery programmatically (rather then using the command line, cause that's annoying)
	app.start(argv=['celery',
					'worker',
					'--loglevel=info',                          # Log level "info"
					'--app=client',                             # App instance
					'--concurrency=1',                          # Limit to 1 worker thread
					'--autoreload',
					'--hostname=worker_{num}'.format(num=num),  # Take the worker name from the command-line
					'-O', 'fair'])                              # Disable prefetching



```

Server:

```
from client import FetchTask
import random

def test():
	while 1:
		x = random.randint(0, 255)
		ret = FetchTask.delay(x)
		val = ret.get()
		print(val)


		# There is a reference leak somewhere, and if the ret and val are not
		# manually deleted, they will persist forever. As such, delete them manually.
		try:
			while 1:
				del ret
		except NameError:
			continue
		try:
			while 1:
				del val
		except NameError:
			continue


		val = None
		ret = None


if __name__ == '__main__':
	test()

```

Basically, the worker has a little bit of persistent state (this is necessary for my application), and it currently just increments a counter with a value passed by the server, and returns the result. For testing, when the value exceeds 1000, a `ValueError` is raised.

The issue is that the propagation of the `ValueError` back to the server seems to have some odd behaviour.

Traceback on the worker:
```
[2015-01-02 20:56:52,927: INFO/MainProcess] Task client.FetchTask[a97c543f-8342-4796-a308-e2d8418fa668] succeeded in 0.04724004294257611s: 807
[2015-01-02 20:56:53,022: INFO/MainProcess] Received task: client.FetchTask[ec7b00a1-65cf-4542-94fc-1943f40f0a30]
[2015-01-02 20:56:53,022: WARNING/Worker-1] x: '195', self.a: '807'
[2015-01-02 20:56:53,094: ERROR/MainProcess] Task client.FetchTask[ec7b00a1-65cf-4542-94fc-1943f40f0a30] raised unexpected: ValueError()
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/local/lib/python3.4/dist-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/media/Storage/Scripts/test/client.py", line 32, in run
    raise ValueError
ValueError

```

Traceback on the server:
```
herp@mainnas:/media/Storage/Scripts/test$ python3 server.py
Task Constructor
Task Constructor
149
332
370
568
807
Traceback (most recent call last):
  File "server.py", line 29, in <module>
    test()
  File "server.py", line 8, in test
    val = ret.get()
  File "/usr/local/lib/python3.4/dist-packages/celery/result.py", line 175, in get
    raise meta['result']
TypeError: exceptions must derive from BaseException

```

I'm not sure if I'm doing something wrong (probably), or what, but I would expect the error to result in either a proper exception from `ret.get()`, or some sort of job failure. Instead, it seems like the internal attempt to raise an exception is doing something wrong. I'm not sure if this is because I'm kind of abusing the `Task` system to get my persistent state, or if there's genuinely a bug in there.

 - Platform: Ubuntu Server 14.04LTS, 64 bit
 - Python 3.4.0
 - Celery version 3.1.17

**Comment by gigimon** [comment]
Yes, we have some problem too. @fake-name maybe You found a problem?

**Comment by fake-name** [comment]
@gigimon - My "solution" was to dump Celery entirely, and write [my own management library](https://github.com/fake-name/AqmpConnector) directly on top of [py-amqp](https://github.com/celery/py-amqp).

Whether that's a proper solution for you depends a lot on what your task load looks like, the library I wrote is basically tailored for my task structure specifically.

**Comment by vesta-crim** [comment]
Anybody else has insights on this ? 

"raise ValueError" or any other type of exception on worker causes "TypeError: exceptions must derive from BaseException" on caller side... 

meta['result'] is a dict instance and Python will not raise a dict instance. I've tried different "propagation" options in Celery but to no avail.

**Comment by thedrow** [comment]
When you raise something you need to raise an instance of an exception.
Try to ```raise ValueError()```.

**Comment by thedrow** [comment]
@vesta-crim ping?

**Comment by vesta-crim** [comment]
@thedrow : Understood, and actually I did raise an instance of an exception even if my "example" did not show it. It seems Celery transforms a raised exception into a dict instance so it can serialize the exception information back to the caller and that the transformation back to an exception instance is not properly done on my setup. I did a workaround by checking a condition before invoking the Celery task but haven't found the proper solution. Regards.

**Comment by thedrow** [comment]
Are you using pickle, json or msgpack as your results serializer?

**Comment by fake-name** [comment]
I think I was using msgpack, fwiw.

**Comment by zvyn** [comment]
I have the same problem in a django setup, with json serializer and redis as back-end (broker+results).
The following trivial example fails:

    """tasks.py"""
    from celery import shared_task

    @shared_task
    def foo(*args):
        raise Exception()

    """views.py"""
    from rest_framework.views import APIView
    from .tasks import bug

    class BugView(APIView):
        def post(self, request, *args, **kwargs):
            bug.delay().get()

result in celery:

    2015-09-14 09:37:20,641 celery.worker.strategy INFO     Received task: project.main.tasks.bug[9574034d-a802-450e-aecc-502a9c2145cf]
    2015-09-14 09:37:20,658 celery.worker.job ERROR    Task project.main.tasks.bug[9574034d-a802-450e-aecc-502a9c2145cf] raised unexpected: Exception()
    Traceback (most recent call last):
       File "/usr/share/project/venv/lib/python3.4/site-packages/celery/app/trace.py", line 240, in trace_task
       R = retval = fun(*args, **kwargs)
       File "/usr/share/project/venv/lib/python3.4/site-packages/celery/app/trace.py", line 438, in __protected_call__
        return self.run(*args, **kwargs)
      File "/root/project/main/tasks.py", line 22, in bug
        raise Exception()
    Exception

result in the server process:

    ceback (most recent call last):
      File "/usr/share/project/venv/lib/python3.4/site-packages/django/core/handlers/base.py", line 132, in get_response
        response = wrapped_callback(request, *callback_args, **callback_kwargs)
      File "/usr/share/project/venv/lib/python3.4/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
        return view_func(*args, **kwargs)
      File "/usr/share/project/venv/lib/python3.4/site-packages/django/views/generic/base.py", line 71, in view
        return self.dispatch(request, *args, **kwargs)
      File "/usr/share/project/venv/lib/python3.4/site-packages/rest_framework/views.py", line 456, in dispatch
        response = self.handle_exception(exc)
      File "/usr/share/project/venv/lib/python3.4/site-packages/rest_framework/views.py", line 453, in dispatch
        response = handler(request, *args, **kwargs)
      File "/vagrant/project/control/views.py", line 64, in post
        return super(ParcelControl, self).post(*args, **kwargs)
      File "/vagrant/project/main/views.py", line 67, in post
        bug.delay().get()
      File "/usr/share/project/venv/lib/python3.4/site-packages/celery/result.py", line 175, in get
        raise meta['result']
    TypeError: exceptions must derive from BaseException

**Comment by aequitas** [comment]
I'm also running into this issue. I resorted to not propagating exceptions and checking the result for the exc_type key, which is hardly a solution.

**Comment by thedrow** [comment]
I'll try to take a look during this week. Ping me if I forget.

**Comment by eliasj** [comment]
Any result?

**Comment by ask** [comment]
Several of these issues have been fixed so unless you have the same traceback, please include it!

**Comment by eliasj** [comment]
Celery config:
```python
    CELERY_BROKER_URL='amqp://dev01:5672//',
    CELERY_RESULT_BACKEND='db+postgresql://eliasj@dev01/node',
    CELERY_IGNORE_RESULT=False,                                                 
    CELERY_RESULT_PERSISTENT=True,                                              
    CELERY_SEND_TASK_ERROR_EMAILS=True,                                         
    CELERY_TRACK_STARTED=True,                                                  
    CELERY_TASK_SERIALIZER='json',                                              
    CELERY_RESULT_SERIALIZER='json',                                            
    CELERY_ACCEPT_CONTENT=['json'], 
```
Task that raises the exception:
```python
@celery.task                                                         
def send_files(command):                                                                             
    args = shlex.split(command)                                             
    result = subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    result.check_returncode()  
```

The traceback for the task:
```
[2016-01-19 06:54:30,839: INFO/MainProcess] Received task: app.sync.send_files[ed86f930-709c-4953-b1a8-9ca6b5c4790b]
[2016-01-19 06:54:32,497: ERROR/MainProcess] Task app.sync.send_files[ed86f930-709c-4953-b1a8-9ca6b5c4790b] raised unexpected: CalledProcessError(23, ['rsync', '-a', '/home/eliasj/bsadasd', 'dev01:/tmp'], b'', b'rsync: link_stat "/home/eliasj/bsadasd" failed: No such file or directory (2)\nrsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1039) [sender=3.0.6]\n')
Traceback (most recent call last):
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/eliasj/Proj/node-copy/app/sync.py", line 69, in send_files
    result.check_returncode()
  File "/usr/local/lib/python3.5/subprocess.py", line 662, in check_returncode
    self.stderr)
subprocess.CalledProcessError: Command '['rsync', '-a', '/home/eliasj/bsadasd', 'dev01:/tmp']' returned non-zero exit status 23
```
The traceback for when I try to get the result:
```python
Traceback (most recent call last):
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/flask_restful/__init__.py", line 477, in wrapper
    resp = resource(*args, **kwargs)
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/flask/views.py", line 84, in view
    return self.dispatch_request(*args, **kwargs)
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/flask_restplus/resource.py", line 42, in dispatch_request
    resp = meth(*args, **kwargs)
  File "/home/eliasj/Proj/node-copy/app/api.py", line 126, in get
    return {'status': get_status(id)}
  File "/home/eliasj/Proj/node-copy/app/sync.py", line 42, in get_status
    result.get(timeout=1)
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/celery/result.py", line 175, in get
    raise meta['result']
TypeError: exceptions must derive from BaseException
```

```python
>>> r = celery.AsyncResult("ed86f930-709c-4953-b1a8-9ca6b5c4790b")
>>> try:
...   r.get(timeout=5)
... except TypeError:
...   pass
...       
>>> r.result
{'exc_type': 'CalledProcessError', 'exc_message': "Command '['rsync', '-a', '/home/eliasj/bsadasd', 'dev01:/tmp']' returned non-zero exit status 23"}
```
Content of the database:
```
node> SELECT * FROM celery_taskmeta WHERE id = 22;
+------+--------------------------------------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+-------------+
|   id | task_id                              | status   | result                                                                                                                                                                                                                                                               | date_done                  | traceback   |
|------+--------------------------------------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+-------------|
|   22 | ed86f930-709c-4953-b1a8-9ca6b5c4790b | FAILURE  | \200\004\225\264\000\000\000\000\000\000\000}\224(\214\010exc_type\224\214\022CalledProcessError\224\214\013exc_message\224\214~Command '['rsync', '-a', '/home/eliasj/bsadasd', 'dev01:/tmp']' returned non-zero exit status 23\224u. | 2016-01-19 05:54:32.493315 | Traceback (most recent call last):
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/eliasj/env/flask/lib/python3.5/site-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/eliasj/Proj/node-copy/app/sync.py", line 69, in send_files
    result.check_returncode()
  File "/usr/local/lib/python3.5/subprocess.py", line 662, in check_returncode
    self.stderr)
subprocess.CalledProcessError: Command '['rsync', '-a', '/home/eliasj/bsadasd', 'dev01:/tmp']' returned non-zero exit status 23             |
+------+--------------------------------------+----------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+-------------+

**Comment by eliasj** [comment]
Hi again, 
I seam to solve my problem. If I change CELERY_RESULT_SERIALIZER to use pickle instead of json it works for me. Is there any drawbacks to use pickle except other programming language then python can't read the result?
```
CELERY_RESULT_SERIALIZER='pickle',
```

**Comment by ask** [comment]
@eliasj: Thanks, it was the sqlalchemy result backend that was the issue here, just fixed it in master and 3.1
